src/oslogin_utils.cc

```diff
@@ -1231,4 +1246,128 @@ bool ContinueSession(bool alt, const string& email, const string& user_token, co
 
   return ret;
 }
+
+static bool ApplyPolicy(const char *user_name, string email, const char *policy, struct AuthOptions opts) {
+  std::stringstream url;
+  url << kMetadataServerUrl << "authorize?email=" << UrlEncode(email) << "&policy=" << policy;
+
+  // Don't try to add fingerprint parameter to policy call if we don't have it.
+  if (opts.fp_len > 0) {
+    url << "&fingerprint=" << opts.fingerprint;
+  }
+
+  string response;
+  long http_code = 0;
+  // Invalid user, just leave from here - the principal will not be allowed/authorized.
+  if (!HttpGet(url.str(), &response, &http_code)) {
+    SysLogErr("Failed to validate organization user %s has login permission.", user_name);
+    return false;
+  }
+
+  if (http_code != 200) {
+    SysLogErr("Failed to validate organization user %s has login permission, "
+              "got HTTP response code: %lu", user_name, http_code);
+    return false;
+  }
+
+  if (!ParseJsonToSuccess(response)) {
+    SysLogErr("Organization user %s does not have login permission.", user_name);
+    return false;
+  }
+
+  return true;
+}
+
+static bool CreateGoogleUserFile(string users_filename) {
+  std::ofstream users_file;
+  users_file.open(users_filename.c_str());
+
+  // OS Login directories are created by another product, guest-agent.
+  // We should be sure a file is opened for writing.
+  if (users_file.is_open()) {
+    // this is only for creating an empty file
+    users_file.close();
+
+    chown(users_filename.c_str(), 0, 0);
+    chmod(users_filename.c_str(), S_IRUSR | S_IWUSR | S_IRGRP);
+    return true;
+  }
+
+  return false;
+}
+
+static bool CreateGoogleSudoersFile(string sudoers_filename, const char *user_name) {
+  std::ofstream sudoers_file;
+  sudoers_file.open(sudoers_filename.c_str());
+
+  // OS Login directories are created by another product, guest-agent.
+  // We should be sure a file is opened for writing.
+  if (sudoers_file.is_open()) {
+    sudoers_file << user_name << " ALL=(ALL) NOPASSWD: ALL\n";
+    sudoers_file.close();
+
+    chown(sudoers_filename.c_str(), 0, 0);
+    chmod(sudoers_filename.c_str(), S_IRUSR | S_IRGRP);
+    return true;
+  }
+
+  return false;
+}
+
+bool AuthorizeUser(const char *user_name, struct AuthOptions opts, string *user_response) {
+  bool users_file_exists, sudoers_exists;
+  string email, users_filename, sudoers_filename;
+  struct stat sudoers_buffer, buffer;
+
+  sudoers_buffer = buffer = { 0 };
+  users_file_exists = sudoers_exists = false;
+
+  if (!ValidateUserName(user_name)) {
+    return false;
+  }
+
+  // Call MDS "users?username=" endpoint.
+  if (!MDSGetUser(user_name, opts.security_key, user_response)) {
+    return false;
+  }
+
+  if (!ParseJsonToEmail(*user_response, &email) || email.empty()) {
+    return false;
+  }
+
+  users_filename = kUsersDir;
+  users_filename.append(user_name);
+  users_file_exists = !stat(users_filename.c_str(), &buffer);
+
+  // Couldn't apply "login" policy for user in question, log it and deny.
```

Comment:
Small nit: these comments explaining what happens in the negative case should go *inside* the block where they are relevant, not outside.

Please apply this across this file; it appears a number of times.

Reply:
Ok, will do. Thanks.

Reply:
Done. Thanks.